fix: 🐛 Added virtual file system support to prevent out of sync

ncarvajalc opened the pull request:
Added virtual file system support to snapshot backend to prevent #current-cloud-backend mismatch with s3 deployment

✅ Closes: aws-amplify/amplify-category-api#88

<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#pull-requests
-->

#### Description of changes

<!--
Thank you for your Pull Request! Please provide a description above and review
the requirements below.
-->

#### Issue #, if available

aws-amplify/amplify-category-api#88

<!-- Also, please reference any associated PRs for documentation updates. -->

#### Description of how you validated changes

`yarn test` and sample app deployment with changes  and no issues when modifying schema.graphql during deployment.

#### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ncarvajalc committed:
fix: 🐛 Added virtual file system support to prevent out of sync

Added virtual file system support to snapshot backend to prevent
#current-cloud-backend mismatch with s3 deployment

✅ Closes: aws-amplify/amplify-category-api#88

alharris-at commented on packages/amplify-provider-awscloudformation/src/push-resources.ts:
```diff
@@ -469,6 +471,48 @@ export const run = async (context: $TSContext, resourceDefinition: $TSObject, re
+  }
+};
+
+const takeBackendSnapshotVFS = (): void => {
```
Given that this is already a pretty bloated file that has a lot of concerns, I'd consider refactoring all the snapshotting and VFS stuff into a standalone utility (will also make the unit tests more clear I think).

alharris-at commented on packages/amplify-cli/src/extensions/amplify-helpers/update-amplify-meta.ts:
```diff
@@ -59,6 +61,21 @@ export const updateAwsMetaFile = (
   return amplifyMeta;
 };
 
+const copyFromVFStoFSSync = (src: string, dest: string): void => {
```
Because this isn't intended to be a general purpose `cp` command, could we revise to accept a single path that we walk recursively?

alharris-at commented on packages/amplify-cli-core/src/state-manager/pathManager.ts:
```diff
@@ -24,6 +24,7 @@ export const PathConstants = {
 
   // resource level
   BuildDirName: 'build',
+  SchemaSnapshot: '.snapshot',
```
As this stands in the `PathConstants`, it looks like it might be reusable, given it's only used in `getBackendSnapshotVFSPath` I might just make inline there so nobody else uses it misunderstanding why it's defined..

alharris-at reviewed (commented):
Unit tests?

sobolk commented on packages/amplify-cli/src/extensions/amplify-helpers/update-amplify-meta.ts:
```diff
@@ -6,6 +7,7 @@ import {
 } from 'amplify-cli-core';
 import { BuildType } from 'amplify-function-plugin-interface';
 import * as fs from 'fs-extra';
+import { fs as vfs } from 'memfs';
```
does this work on windows reliably ?

ncarvajalc replied:
Yes, it works for windows. I tested it yesterday in a windows machine. I will elaborate more in the PR description today

sobolk reviewed (commented):
Can you please elaborate more in the PR description about the problem and solution?
I.e. describe the failure and how this provides recovery from failure, is the recovery automatic?
Could you please elaborate more about data flow before and after the change? I.e. how mutations propagate to local fs, vfs and s3 bucket in happy case and failure case?

Have we considered [s3 versioning](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html) and/or other advanced S3 mechanisms to recover last known good current-cloud-backend ?

ncarvajalc committed:
test: 💍 Added vfs-helper tests and refactored vfs-helpers

-Added unit tests for the new vfs-helpers package\n-Refactored virtual
file system utilities to vfs-helpers

✅ Closes: aws-amplify/amplify-category-api#88

ncarvajalc committed:
test: 💍 Added afterAll function tests

Removed mock files created in tests

ncarvajalc commented:
In response to @sobolk 
As reported in aws-amplify/amplify-category-api#88, (which links to other issues such as: #10118, aws-amplify/amplify-category-api#535) or aws-amplify/amplify-category-api#882, they all reference the same issue which is that there is a mismatch between the local `#current-cloud-backend` resources and the ones deployed. The current workaround to this issue, as suggested by @josefaidt is to follow the following [instructions](https://gist.github.com/josefaidt/be56f2a8a821b084b66e40ce51e20b44):
1. navigate to the s3 console and identify the deployment bucket for our app
2. download `#current-cloud-backend.zip`
3. unzip the contents, correct the GraphQL Schema, and re-zip the folder
	1. be sure to not accidentally zip as a nested folder
	2. `unzip "#current-cloud-backend.zip" -d "#current-cloud-backend"`
	3. `mv "#current-cloud-backend.zip" backup.zip`
	4. correct the schema
	5. `cd "#current-cloud-backend"; zip -r "../#current-cloud-backend.zip" *`
4. re-upload the newly zipped archive to our s3 deployment bucket
5. push from local machine with `amplify push --force`
6. observe successful push 🙌

The mismatch occurs because of the resources being edited while a deployment is in progress. 

## Behavior without the changes
Currently, after an `amplify push`, the following code is executed after all resources have been deployed: 
https://github.com/aws-amplify/amplify-cli/blob/4a4bd618c88239ebb75e728bd265f039f34ee5b3/packages/amplify-provider-awscloudformation/src/push-resources.ts#L344-L367
Notice the `updateamplifyMetaAfterPush` function. This function calls the following:
https://github.com/aws-amplify/amplify-cli/blob/4a4bd618c88239ebb75e728bd265f039f34ee5b3/packages/amplify-cli/src/extensions/amplify-helpers/update-amplify-meta.ts#L263
This is the function in charge of moving the files located in `amplify/backend` to `amplify/#current-cloud-backend`. Notice the directory paths that are being used:
https://github.com/aws-amplify/amplify-cli/blob/4a4bd618c88239ebb75e728bd265f039f34ee5b3/packages/amplify-cli/src/extensions/amplify-helpers/update-amplify-meta.ts#L72-L74
As this uses the `amplify/backend` path, if there was a modification in the backend resources during the deployment process this files would then get copied into the `amplify/#current-cloud-backend` directory, which is then zipped into a file and uploaded to the s3 bucket. This is why the issues are happening, as the `#current-cloud-backend` is not corresponding to the one verified before the deployment process with the actual deployed resources.
## Behavior after the changes
By using a virtual file system (vfs) before the deployment begins it is possible to take a snapshot of the backend just before the deployment process begins. In that case, there is no chance that the developer accidentally changes any of the resource files that are going to get uploaded into the s3 bucket. The developer can actually edit the resources files, but that won't be uploaded as all of the snapshots will reside in the vfs and are going to be retrieved from it.
## Why not s3 versioning?
This is not possible as this is independent from the s3 bucket version, as an actual inaccurate zip is being uploaded to the bucket, with other resources that weren't deployed previously.  

## Data flow before the changes
![Pasted image 20221124112047](https://user-images.githubusercontent.com/66692533/203831077-52fe5ac1-3053-424d-a62d-75e744b6942a.png)

## Data flow after the changes
![Pasted image 20221124112108](https://user-images.githubusercontent.com/66692533/203831090-9a98d668-226e-44f8-8443-a02d7a12b546.png)

sobolk commented:
Thank you @ncarvajalc for the context. It's very helpful.
I like the idea of taking the snapshot. However, I think that if we snapshot into S3 rather than local fs or vfs we might cover more of these kind of cases or at least leave the system in recoverable state (that was my main purpose when I mentioned S3 versioning).

As far as I understand the current cloud backend synchronization happens after push completes. Which makes it prone to local process crashes.
Consider the following scenario.
1. User initiates push
2. backed is copied into vfs
3. CFN deployment kicks off
4. CLI process is interrupted/terminated (think OS crash, VM crash, user sends sigterm or so).
5. CFN deployment completes eventually. (or perhaps process crashes before post push actions trigger).
At this point the ephemeral  snapshot is gone, resources are not in sync and user might have edited or lost backend folder.

My suggestion is to explore how could we use S3 to make this scenario even more resilient and transactional.

ncarvajalc commented:
@sobolk Okay, I can take a look at the S3 approach!

profDevA reviewed (commented):
It looks good

edwardfoyle commented:
Hi @ncarvajalc are you still interested in looking into @sobolk's suggestions?
